src/TestProblem/Hydro/Equilibrium_Cloud/Par_Init_ByFunction_Equilibrium_Cloud.cpp

```diff
@@ -0,0 +1,163 @@
+//-------------------------------------------------------------------------------------------------------
+
+void Par_Init_ByFunction_Equilibrium_Cloud( const long NPar_ThisRank, const long NPar_AllRank,
+                                  real *ParMass, real *ParPosX, real *ParPosY, real *ParPosZ,
+                                  real *ParVelX, real *ParVelY, real *ParVelZ, real *ParTime,
+                                  real *AllAttribute[PAR_NATT_TOTAL] )
+{
+   // Define particles' attributes array
+   real *Mass_AllRank   = NULL;
+   real *Pos_AllRank[3] = { NULL, NULL, NULL };
+   real *Vel_AllRank[3] = { NULL, NULL, NULL };
+
+   // Define the Particle IC Constructor
+   Particle_IC_Constructor Filename_Loader;
+
+   if ( MPI_Rank == 0 )    Aux_Message( stdout, "%s ...\n", __FUNCTION__ );
+
+   // only the master rank will construct the initial condition
+   if ( MPI_Rank == 0 ){
+
+      // Allocate memory for particles' attributes array
+      Mass_AllRank = new real [NPar_AllRank];
+      for (int d=0; d<3; d++)
+      {
+         Pos_AllRank[d] = new real [NPar_AllRank];
+         Vel_AllRank[d] = new real [NPar_AllRank];
+      }
+
+      // Input filenames as parameters into Filename_Loader
+      Filename_Loader.Read_Filenames("Input__TestProb");
+      int Par_Idx_Last = 0;
+      double Par_Ratio = 0.0;
+
+      for(int k=0;k<Filename_Loader.filenames.Cloud_Num;k++){
+         // initialize Particle_IC_Constructor for each cloud
+         Particle_IC_Constructor Cloud_Constructor;
+         Cloud_Constructor.Load_Physical_Params(Filename_Loader.filenames,k);
+         Cloud_Constructor.Init();
+
+         // calculate particle number for each cloud
+         if((Par_Ratio + Cloud_Constructor.params.Cloud_Par_Num_Ratio) > 1.0){
+            Aux_Error( ERROR_INFO, "The sum of particle number ratios of all clouds exceeds 1!! Please check!");
+         }
+         if((k==Filename_Loader.filenames.Cloud_Num-1)&&((Par_Ratio + Cloud_Constructor.params.Cloud_Par_Num_Ratio) != 1.0)){
+            Aux_Error( ERROR_INFO, "The sum of particle number ratios of all clouds is not equal to 1.0!! Please check!");
+         }
+         int Par_Idx_Start = Par_Idx_Last;
+         Par_Idx_Last += Cloud_Constructor.params.Cloud_Par_Num_Ratio*NPar_AllRank;
+         Par_Ratio += Cloud_Constructor.params.Cloud_Par_Num_Ratio;
+         if(k==Filename_Loader.filenames.Cloud_Num-1){
+            Par_Idx_Last = NPar_AllRank;
+         }
+
+         // set equilibrium initial conditions for each cloud
+         // Input Mass_AllRank, Pos_AllRank, Vel_AllRank into this function for particles' masses, positions, and velocities.
+         // NPar_AllRank is the total number (including all clouds) of particles.
+         // Par_Idx_Start is the starting index for the particle in this cloud, and Par_Idx_Last is the last index.
+         Cloud_Constructor.Par_SetEquilibriumIC(Mass_AllRank, Pos_AllRank, Vel_AllRank,NPar_AllRank,Par_Idx_Start,Par_Idx_Last);
```

Comment:
Revise it according to the simplified parameters of `Par_SetEquilibriumIC()` that I suggest in `Par_EquilibriumIC.cpp`.